File: 프로그래머스/2/181188. 요격 시스템/요격 시스템.java
import java.util.PriorityQueue;

public class Solution {

    public static int solution(int[][] targets) {
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> {
            return a[0] - b[0];
        });

        for (int[] target : targets) {
            pq.add(target);
        }

        int answer = 0;

        int end = -1;

        while (!pq.isEmpty()) {
            int[] cur = pq.poll();

            if (cur[0] >= end) {
                answer++;
                end = cur[1];
            } else {
                end = Math.min(end, cur[1]);
            }
        }
        return answer;
    }
}

File: 프로그래머스/2/181188. 요격 시스템/요격 시스템.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <cmath>
using namespace std;

int solution(vector<vector<int>> targets) {
    int answer = 0;
    
    sort(targets.begin(), targets.end(), [](vector<int> a, vector<int> b){
        if (a[0] == b[0])
            return a[1] < b[1];
        
        return a[0] < b[0];
    });
    
    int start = -1;
    int end = -1;
    for(int i = 0; i < targets.size(); ++i)
    {
        if (targets[i][0] >= end)
        {
            start = targets[i][0];
            end = targets[i][1];
            ++answer;
            continue;
        }
        else
        {
            end = min(targets[i][1], end);
        }
    }
    
    return answer;
}